### ISSUE TITLE
[no-invalid-this] Add improved `no-invalid-this` rule that allow `this` access in class field arrow functions

### ISSUE THREAD

**Issue by josundt** [typescript-eslint/typescript-eslint#1571]
**Repro**
The ESLint _standard_ rule **`'no-invalid-this'`**, gives an error when accessing **`this`** inside a class field arrow function (see repro code below).

**babel/eslint** proviides its own rule `'babel/no-invalid-this'` that addresses this problem.
Reference: _[Fixing eslint no-invalid-this error for fat arrow class methods](https://medium.com/@chestozo/fixing-eslint-no-invalid-this-error-for-fat-arrow-class-methods-a56908ca8bb6)_

I therefore request that **@typescript/eslint** adds its own improved **`'no-invalid-this'`** rule, the same way that babel-eslint has done; to address this (and potential other) problems with the standard ESLint **`'no-invalid-this'`** rule.

**Repro rules:**
```javascript
{
  "rules": {
    // Current rule that causes problems:
    "no-invalid-this": "error"
    // My requested replacement rule which will address the problem:
    // "@typescript-eslint/no-invalid-this": "error"
  }
}
```

**Repro code:**
```typescript
type EventHandler<TEvent extends Event> = (event: TEvent) => any;

export class KeyboardDebugLogger {
    constructor(private readonly logger: Console) {}

    private readonly onDocumentKeyDown: EventHandler<KeyboardEvent> = event => {
        // Next line gives error for ESLint standard rule "no-invalid-this"
        this.logger.debug('keydown event triggered');
    };

    startListening(): void {
        document.addEventListener('keydown', this.onDocumentKeyDown, true);
    }

    stopListening(): void {
        document.removeEventListener('keydown', this.onDocumentKeyDown, true);
    }
}
``````

**Expected Result**
No error

**Actual Result**
Error

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.19.0` |
| `@typescript-eslint/parser`        | `2.19.0` |
| `TypeScript`                       | `3.7.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.14.0` |
| `npm`                              | `6.13.4` |

**Comment by glen-84** [comment]
Duplicate of #491.
